plugins/main/tap/tap.go

```diff
@@ -0,0 +1,464 @@
+	"github.com/containernetworking/plugins/pkg/ipam"
+	"github.com/containernetworking/plugins/pkg/ns"
+	bv "github.com/containernetworking/plugins/pkg/utils/buildversion"
+	"github.com/containernetworking/plugins/pkg/utils/sysctl"
+)
+
+type NetConf struct {
+	types.NetConf
+	MultiQueue bool   `json:"multiQueue"`
+	MTU        int    `json:"mtu"`
+	Mac        string `json:"mac,omitempty"`
+
+	RuntimeConfig struct {
+		Mac string `json:"mac,omitempty"`
+	} `json:"runtimeConfig,omitempty"`
+}
+
+// MacEnvArgs represents CNI_ARG
+type MacEnvArgs struct {
+	types.CommonArgs
+	MAC types.UnmarshallableString `json:"mac,omitempty"`
+}
+
+func init() {
+	// this ensures that main runs only on main thread (thread group leader).
+	// since namespace ops (unshare, setns) are done for a single thread, we
+	// must ensure that the goroutine does not jump from OS thread to thread
+	runtime.LockOSThread()
+}
+
+func loadConf(args *skel.CmdArgs) (*NetConf, string, error) {
+	n := &NetConf{}
+	if err := json.Unmarshal(args.StdinData, n); err != nil {
+		return nil, "", fmt.Errorf("failed to load netconf: %v", err)
+	}
+	if args.Args != "" {
+		e := MacEnvArgs{}
+		err := types.LoadArgs(args.Args, &e)
+		if err != nil {
+			return nil, "", err
+		}
+
+		if e.MAC != "" {
+			n.Mac = string(e.MAC)
+		}
+	}
+
+	if n.RuntimeConfig.Mac != "" {
+		n.Mac = n.RuntimeConfig.Mac
+	}
+
+	return n, n.CNIVersion, nil
+}
+
+func CloseOnExec(fd int) {
+	syscall.CloseOnExec(fd)
+}
+
+func setContainerSeBool() error {
+	output, err := exec.Command("cat", "/sys/fs/selinux/booleans/container_use_devices").CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed to run getsebool command %s: %v", string(output), err)
+	}
+	if strings.Contains(string(output), "off") {
+		output, err := exec.Command("setsebool", "-P", "container_use_devices", "true").CombinedOutput()
+		if err != nil {
+			return fmt.Errorf("failed to run setsebool command %s: %v", string(output), err)
+		}
+	}
+	return nil
+}
+
+func createSelinuxTap(conf *NetConf, tmpName string) error {
+	if err := selinux.SetExecLabel("system_u:system_r:container_t:s0"); err != nil {
+		return fmt.Errorf("failed set socket label: %v", err)
+	}
+
+	minFDToCloseOnExec := 3
+	maxFDToCloseOnExec := 256
+	// we want to share the parent process std{in|out|err} - fds 0 through 2.
+	// Since the FDs are inherited on fork / exec, we close on exec all others.
+	for fd := minFDToCloseOnExec; fd < maxFDToCloseOnExec; fd++ {
+		CloseOnExec(fd)
+	}
+
+	tapDeviceArgs := []string{"tuntap", "add", "mode", "tap", "name", tmpName}
+	if conf.MultiQueue {
+		tapDeviceArgs = append(tapDeviceArgs, "multi_queue")
+	}
+
+	output, err := exec.Command("ip", tapDeviceArgs...).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed to run command %s: %v", output, err)
+	}
+
+	if err := selinux.SetExecLabel("system_u:system_r:container_t:s0"); err != nil {
+		return fmt.Errorf("failed set socket label: %v", err)
+	}
```

Comment:
isn't this duplicated ?

Reply:
I don't think it duplicated we need this to run the second time the ip link command to set the MTU and mac address if needed